include/h2o.h

```diff
@@ -1855,6 +1855,11 @@ void h2o_headers_register_configurator(h2o_globalconf_t *conf);
 
 /* lib/proxy.c */
 
+/**
+ * A callback to be able to override the destination url on a per-request basis
+ */
+typedef int (*h2o_proxy_get_upstream)(h2o_handler_t *, h2o_req_t *, h2o_url_t **, h2o_socketpool_t **, void *ctx);
```

Comment:
How about making it a callback object instead of a callback that takes a void pointer? Something like below.

```
typedef struct st_h2o_proxy_select_upstream_t {
    int (*select_upstream)(struct st_h2o_proxy_self_upstream_t *self, h2o_req_t *, h2o_url_t *, h2o_socketpool_t *);
} h2o_proxy_select_upstream_t;
```

Doing that way, only one pointer (pointing to `h2o_proxy_select_upstream_t`) needs to be added to `h2o_proxy_config_vars_t`. Every application that instantiates the object can allocate extra memory behind `h2o_proxy_select_upstream_t` and use that space to store its own data.